[PyTorch] Enable lightweight dispatch as an option in cmake build

larryliu0820 opened the pull request:
Adding a new cmake option `USE_LIGHTWEIGHT_DISPATCH`. This option needs to be enabled along with `STATIC_DISPATCH_BACKEND` option.

In `Codegen.cmake` generate the code for codegen unboxing and add it into the build in `src/ATen/CMakeLists.txt`.

This PR is stacked on top of #69881 in `static_dispatch` branch.

larryliu0820 committed:
[PyTorch] Add codegen unboxing ability (#69881)

Summary:
This PR (re)introduces python codegen for unboxing wrappers. Given an entry of `native_functions.yaml` the codegen should be able to generate the corresponding C++ code to convert ivalues from the stack to their proper types. To trigger the codegen, run
```
tools/jit/gen_unboxing.py -d cg/torch/share/ATen --enable_codegen_unboxing
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/69881

Test Plan:
See code is generated using
```
tools/jit/gen_unboxing.py -d cg/torch/share/ATen --enable_codegen_unboxing
```
Will add unit test and CI in later diffs.

Differential Revision: D33692299

Pulled By: larryliu0820

fbshipit-source-id: 731a50befa34a1fd1533ed9b6f518adbc41f71cd

larryliu0820 committed:
[PyTorch] Enable lightweight dispatch as an option in cmake build

larryliu0820 committed:
[PyTorch] Enable lightweight dispatch as an option in cmake build

larryliu0820 committed:
[PyTorch] Enable lightweight dispatch as an option in cmake build

larryliu0820 committed:
[PyTorch] Refactor make_file_manager so that it can be reused between gen.py and gen_unboxing.py
Also modified docs for `unboxing.py` to better reflect the benefit of having codegen unboxing.

larryliu0820 committed:
[PyTorch] Fix lint

larryliu0820 committed:
[PyTorch] Fix lint

larryliu0820 committed:
[PyTorch] Remove code to generate TensorOptions
Pass all the values directly to C++ API

larryliu0820 committed:
[PyTorch] Remove unused imports

larryliu0820 committed:
[PyTorch] Refactor using translate to generate argument list

larryliu0820 committed:
[PyTorch] Fix lint

larryliu0820 committed:
[PyTorch] Optimize import

larryliu0820 committed:
[PyTorch] Address comments, mostly refactoring

larryliu0820 committed:
[PyTorch] Fix lint

larryliu0820 committed:
[PyTorch] Fix lint again

larryliu0820 committed:
[PyTorch] Fix unused import

larryliu0820 committed:
[PyTorch] Rename function name and modify a few comments

larryliu0820 committed:
[PyTorch] Add support for bazel build

larryliu0820 committed:
[PyTorch] Fix BUILD.bazel

larryliu0820 committed:
[PyTorch] Fix bazel

larryliu0820 committed:
[PyTorch] Retry fixing BUILD.bazel

larryliu0820 committed:
[PyTorch] Generate binding list instead of expr list for arguments

larryliu0820 committed:
[PyTorch] Address part of the comments

larryliu0820 committed:
[PyTorch] Address the rest of comments

larryliu0820 committed:
[PyTorch] Add return type in function signature

larryliu0820 committed:
[PyTorch] Remove unused imports

larryliu0820 committed:
[PyTorch] Address comments

larryliu0820 committed:
[PyTorch] Fix lint in unboxing.py

larryliu0820 committed:
[PyTorch] Remove unused import

larryliu0820 committed:
[PyTorch] Remove merge conflict

larryliu0820 committed:
[PyTorch] Add trailing newline

larryliu0820 committed:
[PyTorch] Address comments inline cpp type str, cleanup comments, put arguments into one line

larryliu0820 committed:
[PyTorch] Add space between arguments

iseeyuan commented on scripts/build_ios.sh:
```diff
@@ -88,6 +88,12 @@ if [ "${TRACING_BASED}" == 1 ]; then
 fi
+if [ "${USE_LIGHTWEIGHT_DISPATCH}" == 1 ]; then
+  CMAKE_ARGS+=("-DUSE_LIGHTWEIGHT_DISPATCH=ON")
+  CMAKE_ARGS+=("-DSTATIC_DISPATCH_BACKEND=CPU")
```
I'm wondering if there's actual use case with STATIC_DISPATCH_BACKEND other than CPU? Even with USE_LIGHTWEIGHT_DISPATCH on, there could be multiple backends that need to be code-gen, like CPU, QuantizedCPU. Any reason there are two switches of USE_LIGHTWEIGHT_DISPATCH and TATIC_DISPATCH_BACKEND?

larryliu0820 replied:
So `USE_LIGHTWEIGHT_DISPATCH` not only generate kernels with static dispatch backend(s), but also generate C++ code to register them into JIT op registry. I think it makes sense to keep them separated.

> there could be multiple backends that need to be code-gen, like CPU, QuantizedCPU
Yep we will support multiple backends for static dispatch as next step.

I guess it makes sense to not bundle `USE_LIGHTWEIGHT_DISPATCH` and `STATIC_DISPATCH_BACKEND` together. I put it this way just for testing purpose, will change when I publish it.

iseeyuan replied:
I see, so 

USE_LIGHTWEIGHT_DISPATCH:  STATIC_DISPATCH_BACKEND must be set + JIT op registry.

STATIC_DISPATCH_BACKEND: specify which backend to statically dispatch to. In future, it may be extended to support multiple backends.

It would be great to have the logic in the PR summary and also in the comments.

larryliu0820 committed:
[PyTorch] Incorporate #71504 changes

Avoid registering into dispatcher but only register into JIT op registry. This commit enables the CI linux-xenial-py3-clang5-mobile-custom-build-static test with lightweight dispatch enabled.

larryliu0820 committed:
[PyTorch] Fix codegen.cmake

larryliu0820 committed:
[PyTorch] Fix codegen.cmake by setting sources conditionally

iseeyuan reviewed (approved):
LGTM.

larryliu0820 committed:
[PyTorch] Modify static dispatch CI test to support codegen unboxing

larryliu0820 committed:
[PyTorch] Fix cmake lint

larryliu0820 committed:
[PyTorch] Enable both mobile and server build CI tests

larryliu0820 committed:
[PyTorch] Too many jobs failing, reverting Bazel file changes to debug

larryliu0820 committed:
[PyTorch] Fix lint

larryliu0820 committed:
Merge branch 'static_dispatch' into larryliu-scratch

larryliu0820 commented on test/mobile/custom_build/lite_predictor.cpp:
```diff
@@ -0,0 +1,41 @@
+  auto module = loadModel(argv[1]);
+  auto input = torch::ones({1, 3, 224, 224});
+  auto output = [&]() {
+    return module.forward({input}).toTensor();
```
Yes good call out. This model only contains a dozen operators and the test will only verify the correctness of these operators with lightweight dispatch. Like you said I'm going to increase test coverage by adding unit tests similar to [test_lite_interpreter.cpp](https://github.com/pytorch/pytorch/blob/master/test/cpp/jit/test_lite_interpreter.cpp), probably in another PR.
